File: Irrgarten_Java/src/irrgarten/CombatElement.java
package irrgarten;

/**
 * Clase que representa los elementos de combate que utiliza el jugador durante 
 * los combates a lo largo del juego.
 * 
 * @author Arturo Olivares Martos
 * @author Joaquín Avilés de la Fuente
 */

// TODO: poner visibilidad delante?
abstract class CombatElement {
    /**
     * Efecto del elemento de combate.
     */
    private float effect;
    
    /**
     * Número de usos disponibles del elemento de combate.
     */
    private int uses;
    
    /**
     * Constructor de la clase abstracta CombatElement
     * @param effect Efecto del elemento de combate
     * @param uses El número de usos disponibles del elemento de combate
     */
    public CombatElement (float effect, int uses){
        this.effect=effect;
        this.uses=uses;
    }
    
    /**
     * Método para realizar el efecto correspondiente al elemento de combate,
     * se reduce en uno su uso.
     * @return La intensidad del efecto de dicho elemento de combate.
     */
    protected float produceEffect(){
        if(uses>0){
            uses--;
            return effect;
        }
        else
            return 0;
    }
    
    /**
     * Método que indica si se descartará el elemento de combate en función de
     * sus usos
     * @return devuelve true o false si se descarta o no
     */
    public boolean discard(){
        return Dice.discardElement(this.uses);
    }
    
    /**
     * Método que devuelve una representación en forma de cadena de caracteres
     * del estado interno del elemento de combate.
     * @return Representación en forma de cadena de caracteres del estado interno del elemento de combate.
     */
    @Override
    public String toString(){
        return "[" + effect + ", " + uses + "]";
    }
}

File: Irrgarten_Ruby/combat_element.rb
# encoding: UTF-8

require_relative 'dice'


module Irrgarten

    # Esta clase representa un elemento de combate, que puede ser un arma ({Weapon}) o un escudo ({Shield}).
    #
    # @abstract No se debe instanciar directamente.
    #
    # @author Joaquin Avilés de la Fuente
    # @author Arturo Olivares Martos
    class CombatElement

        # Constructor de la clase
        #
        # @param effect [float] efecto del elemento de combate
        # @param uses [int] usos que tiene el escudo
        def initialize(effect, uses)
            @effect = effect.to_f
            @uses = uses.to_i
        end

        protected
        # Método que produce el efecto del elemento de combate y decrementa el número de usos en uno.
        #
        # @return [float] Si tiene algún uso disponible devuelve el uso del elemento. En caso contrario (**uses**==**0**) devuelve 0.
        def produce_effect
            if @uses > 0
                @uses -= 1
                return @effect
            else
                return 0.0
            end
        end

        public

        # Método que indica si se descartará el elemento de combate.
        #
        # @see Dice#dicard_element
        # @return [boolean] devuelve **true** o **false** si se descarta o no
        def discard
            return Dice.dicard_element(@uses)
        end

        # Método que muestra en una cadena el estado del elemento, en cuanto a
        # uso y efecto.
        #
        # @return [String] devuelve una cadena que muestra usos y efecto del elemento
        def to_s
            return "[#{@effect}, #{@uses}]"
        end


    end # class
end # module
